ingestion/src/legacy/checked.rs

```diff
@@ -0,0 +1,369 @@
+    // TODO: Louise directly specified topic partitions 0 and 1 to subscribe to. Was there a reason
+    // for this? The kafka group coordinator should automatically assign partitions to consumers
+    // such that the group covers all partitions, and we shouldn't have to worry about it. On that
+    // note though, should be spawn a consumer task for each partition? It should increase our
+    // throughput
+    let consumer = create_consumer(brokers.as_str(), group.as_str(), topic);
+
+    // Channel buffer size here is based on pure vibes, feel free to change it
+    let (parse_tx, mut parse_rx) = tokio::sync::mpsc::channel::<(CheckedMsg, Offset)>(1);
+    let (db_tx, mut db_rx) =
+        tokio::sync::mpsc::channel::<(Vec<UnlabelledDatum>, Offset)>(DB_BUFFER_SIZE);
+    let (offset_tx, mut offset_rx) = tokio::sync::mpsc::channel::<Offset>(1);
+
+    // Needs to be on a sync thread because processing a message is sync and I measured it to take
+    // ~200us on average. Tokio tasks should not go more than 10-100us between await points
+    // according to tokio devs to avoid choking the runtime. See:
+    // https://ryhl.io/blog/async-what-is-blocking/
+    let _parse_thread = std::thread::spawn(move || {
+        while let Some((message, offset)) = parse_rx.blocking_recv() {
+            let raw_data = match parse_message(&message) {
+                Ok(raw_data) => raw_data,
+                Err(e) => {
+                    metrics::counter!(KAFKA_CHECKED_FAILURES).increment(1);
+                    error!("Failed to parse kafka message: {}, message: {}", e, message,);
+                    continue;
+                }
+            };
+            if let Err(e) = db_tx.blocking_send((raw_data, offset)) {
+                metrics::counter!(KAFKA_CHECKED_FAILURES).increment(1);
+                error!("Failed to send parsed kafka message to db task: {}", e);
+                break;
+            };
+        }
+    });
+    let db_task = tokio::task::spawn(async move {
+        let mut open_conn = pools
+            .open
+            .get()
+            .await
+            .expect("legacy::checked DB task could'nt connect to open DB");
+        let mut restricted_conn = pools
+            .restricted
+            .get()
+            .await
+            .expect("legacy::checked DB task could'nt connect to restricted DB");
+
+        let open_query = open_conn
+            .prepare(QUERY_STR)
+            .await
+            .expect("legacy::checked DB task couldn't prepare open query");
+        let restricted_query = restricted_conn
+            .prepare(QUERY_STR)
+            .await
+            .expect("legacy::checked DB task couldn't prepare restricted query");
+
+        let mut raw_buffer: Vec<(Vec<UnlabelledDatum>, Offset)> =
+            Vec::with_capacity(DB_BUFFER_SIZE);
+
+        while db_rx.recv_many(&mut raw_buffer, DB_BUFFER_SIZE).await != 0 {
+            let offset = raw_buffer.last().unwrap().1.clone();
```

Comment:
Are we sure there is always a last? if it can be empty this is unsafe?

Reply:
Yes, from the [tokio docs](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.recv_many):

> For non-zero values of limit, this method will never return 0 unless the channel has been closed and there are no remaining messages in the channel’s queue. This indicates that no further values can ever be received from this Receiver. The channel is closed when all senders have been dropped, or when [close](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.close) is called.

Note that on the line above, we check `!= 0` to exit the loop in the case where the channel is closed